src/node_file.cc

```diff
@@ -1095,6 +1099,46 @@ static void ExistsSync(const FunctionCallbackInfo<Value>& args) {
   args.GetReturnValue().Set(err == 0);
 }
 
+bool FastExistsSync(v8::Local<v8::Object> recv,
+                    const FastOneByteString& string,
+                    // NOLINTNEXTLINE(runtime/references) This is V8 api.
+                    FastApiCallbackOptions& options) {
+  Environment* env = Environment::GetCurrent(recv->GetCreationContextChecked());
+
+  MaybeStackBuffer<char> path;
+
+  path.AllocateSufficientStorage(string.length + 1);
+  memcpy(path.out(), string.data, string.length);
+  path.SetLengthAndZeroTerminate(string.length);
+
+  if (UNLIKELY(!env->permission()->is_granted(
+          permission::PermissionScope::kFileSystemRead, path.ToStringView()))) {
+    options.fallback = true;
+    return false;
+  }
+
+  uv_fs_t req;
+  auto make = OnScopeLeave([&req]() { uv_fs_req_cleanup(&req); });
+  FS_SYNC_TRACE_BEGIN(access);
+  int err = uv_fs_access(nullptr, &req, path.out(), 0, nullptr);
+  FS_SYNC_TRACE_END(access);
+
+#ifdef _WIN32
+  // In case of an invalid symlink, `uv_fs_access` on win32
+  // will **not** return an error and is therefore not enough.
+  // Double check with `uv_fs_stat()`.
+  if (err == 0) {
+    FS_SYNC_TRACE_BEGIN(stat);
+    err = uv_fs_stat(nullptr, &req, path.out(), nullptr);
+    FS_SYNC_TRACE_END(stat);
+  }
+#endif  // _WIN32
```

Comment:
The fs operation part can be wrapped in a wrapper and shared with the slow callback to avoid getting out of sync.